### PR TITLE
feat: RunnerSets

### PR DESCRIPTION
**Pull request by mumoshu**
~~This is still WIP but I just wanted to submit this as a draft for the early review :)~~

~~The build passes and the implementation should be complete. But it's not tested at all. I'll make this out of the draft state once I've tested it on a real cluster.~~

UPDATE: This is tested E2E with running a Ginkgo suite on GitHub Actions. See `make test` and `pkg/controllers/runnerset_controller_test.go`.

---

RunnerSet is basically ReplicaSet for Runners.

It is responsible for maintaining number of runners to match the desired one. That is, it creates missing runners from `.Spec.Template` and deletes redundant runners.

Similar to ReplicaSet, this does not support rolling update of runners on its own. We might want to later add `RunnerDeployment` for that. But that's another story.

Notes:

- RunnerSet has `availableReplicas` and `readyReplicas` under `Status`, which are showns in `kubectl get` output as `AVAILABLE` and `READY` respectively. These are borrowed from ReplicaSet.

### CODE REVIEW

**Review by summerwind — COMMENTED** [review]
Thank you for the contribution! This change makes Runner easier to scale.
Can you please update CRD and RBAC with `make manifests`?

File: api/v1alpha1/runner_types.go [hunk 1]
```diff
@@ -88,6 +88,33 @@ type RunnerList struct {
...
+	Replicas int `json:"replicas"`
+
+	Template RunnerSpec `json:"template"`
+}
+
+type RunnerSetStatus struct {
+	AvailableReplicas int `json:"availableReplicas"`
+	ReadyReplicas     int `json:"readyReplicas"`
+}
+
```
[the remark this answers is not in the document]
Comment (mumoshu) [reply api/v1alpha1/runner_types.go]: @summerwind Good catch! And I think we do need it now, as I've observed while testing this, listing resources doesn't work without it.

**Comment by mumoshu** [comment]
@summerwind I believe this is now ready for your review.

Not sure this can either be a long-term solution or not. But I think this still worth being included in this project as the baseline to compare future solutions against.

Background: As we've talked face-to-face, a StatefulSe-and-mutating-webhook based solution may replace this. We may completely remove RunnerSet and instead just introduce RunnerDeployment based on that.

**Comment by mumoshu** [comment]
I've added `make tests` to run E2E tests on GitHub Actions. I've confirmed it to work on my fork. Please see `Run tests` job in https://github.com/mumoshu/actions-runner-controller-ci/runs/459416632?check_suite_focus=true

**Comment by summerwind** [comment]
@mumoshu Thank you for adding tests in addition to the implementation!
I agree with using this as a "baseline" implementation.

The code looks good to me. Let's merge.

**Review by summerwind — APPROVED** [review]

**Comment by TomaszKrz** [comment]
ignore

**Comment by TomaszKrz** [comment]
please ignore previous accidental automation comment; no action required.

**Pull request merged**